### PR TITLE
Add SILOptimizer/Utils/Differentiation directory.

### PR DESCRIPTION
**Pull request by bgogul**
Related to https://bugs.swift.org/browse/TF-993

### CODE REVIEW

**Comment by bgogul** [comment]
@swift-ci please test tensorflow

**Review by dan-zheng — APPROVED** [review]

File: include/swift/SILOptimizer/Utils/Differentiation/Common.h [hunk 1]
```diff
@@ -1,4 +1,4 @@
-//===--- Differentiation.h - SIL Automatic Differentiation ----*- C++ -*---===//
+//===--- Common.h - Automatic Differentiation Common Utils ----*- C++ -*---===//
```
Comment (rxwei) [inline include/swift/SILOptimizer/Utils/Differentiation/Common.h]: What makes derivative lookup not "common"?

Comment (bgogul) [reply include/swift/SILOptimizer/Utils/Differentiation/Common.h]: `Common.h` is just a temporary catch all file. I plan to add more files to this directory. Eventually, code in `Common.h` will get refactored out into other files and will get removed.

**Comment by bgogul** [comment]
@swift-ci please test tensorflow

**Review by rxwei — APPROVED** [review]

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Pull request merged**